kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+  name = lower(name)
+
+  if type(versions) ~= "table" then
+    error("invalid versions array for service " .. name)
+  end
+
+  local supported_service = supported_services[name]
+  if not supported_service then
+    return { description = "unknown service." }
+  end
+
+  return negotiate_version(name, versions, supported_service)
+end
+
+local function log_negotiation_result(name, version)
+  if version.version ~= nil then
+    log(NOTICE, "service accepted: \"", name, "\", version: ", version.version, ", description: ", version.description)
+
+  else
+    log(NOTICE, "service rejected: \"", name, "\", reason: ", version.description)
+  end
+end
+
+local function negotiate_services(services_requested)
+  local services = {}
+
+  for name, versions in pairs(services_requested) do
+    if type(name) ~= "string" or type(versions) ~= "table" then
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
+
+    log_negotiation_result(name, negotiated_version)
+  end
+
+  return services
+end
+
+
+local function register_client(cluster_data_plane_purge_delay, id, client_node)
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = id, }, {
+    last_seen = time(),
+    config_hash = DECLARATIVE_EMPTY_CONFIG_HASH,
+    hostname = client_node.hostname,
+    ip = var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+  }, { ttl = cluster_data_plane_purge_delay })
+
+  if not ok then
+    log(ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return error(err)
+  end
+end
+
+function _M.init_negotiation_server(service, conf)
+  service:import("kong.services.negotiation.v1.negotiation")
+  service:set_handler("NegotiationService.NegotiateServices", function(peer, nego_req)
+    local ok, result = pcall(function()
+      local dp_id = peer.id
+      log(NOTICE, "negotiating services for DP: ", dp_id)
+      local services_requested = unwrap_services(nego_req.services_requested)
+      verify_request(nego_req)
+
+      nego_req.node.sync_status = verify_node_compatibility(nego_req.node)
+      local services = negotiate_services(services_requested)
+      register_client(conf.cluster_data_plane_purge_delay, dp_id, nego_req.node)
+
+      local nego_result = {
+        services = services
+      }
+
+      return nego_result
+    end)
+
+    if not ok then
+      log(ERR, _log_prefix, result)
+      return { error_message = result }
+    end
+
+    return result
+  end)
+end
+
+-- TODO: use event to notify other workers
+-- Currently we assume only worker 0 cares about wRPC services
+local negotiated_service
+local function init_negotiated_service_tab()
+  if not negotiated_service then
+    negotiated_service = {}
+
+  else
+    table_clear(negotiated_service)
+  end
+end
+
+local function set_negotiated_service(name, verison)
```

Comment:
This function is too small, I think it is not worthy.

Reply:
The reason to have this function is not to reduce the code size, but keep the abstraction.

Reply:
it's worthy to create variables and functions just to put names on things.  function calls are basically free (and invisible to the tracer)